### PR TITLE
Add some fast CRC16 algorithms implementation

### PR DESCRIPTION
**Pull request by 0xA50C1A1**
Please sign (check) the below before submitting the Pull Request:

- [x] I have signed the ntop Contributor License Agreement at https://github.com/ntop/legal/blob/main/individual-contributor-licence-agreement.md
- [x] I have read the contributing guide lines at https://github.com/ntop/nDPI/blob/dev/CONTRIBUTING.md
- [x] I have updated the documentation (in doc/) to reflect the changes made (if applicable)

Link to the related issue: #2194 

Describe changes:

Implementation of some commonly used CRC16 algorithms. Should make it easier to implement dissectors of protocols such as IEEE C37.118, Ether-S-Bus and Genisys.

### CODE REVIEW

**Comment by 0xA50C1A1** [comment]
@IvanNardi @utoni could you take a look, please?

File: src/lib/third_party/src/ndpi_crc.c [hunk 1]
```diff
@@ -0,0 +1,97 @@
+#include "ndpi_main.h"
+#include "ndpi_crc.h"
+
+static const u_int16_t crc16_ccitt_table[256] = {
```
Comment (utoni) [inline src/lib/third_party/src/ndpi_crc.c]: Do you have any sources/references for those lookup tables?

Comment (0xA50C1A1) [reply src/lib/third_party/src/ndpi_crc.c]: > Do you have any sources/references for those lookup tables?

It's just a CRC for each byte from 0 to 255, the polynomials are 0x1021 and 0x8408.

File: src/lib/third_party/include/ndpi_crc.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#ifndef _NDPI_CRC_
+#define _NDPI_CRC_
+
+u_int16_t ndpi_crc16_ccit(const void* data, size_t n_bytes);
+u_int16_t ndpi_crc16_ccit_false(const void *data, size_t n_bytes);
+u_int16_t ndpi_crc16_xmodem(const void *data, size_t n_bytes);
+
+#endif
```
Comment (IvanNardi) [inline src/lib/third_party/include/ndpi_crc.h]: Could you avoid adding a new header and move these declaration in `src/include/ndpi_api.h`, near the similar crc32 ones, please?

File: src/lib/third_party/src/ndpi_crc.c [hunk 3]
```diff
@@ -0,0 +1,97 @@
+#include "ndpi_main.h"
```
Comment (IvanNardi) [inline src/lib/third_party/src/ndpi_crc.c]: The same here. Could you move this code in `src/lib/ndpi_analyze.c`, near the crc32 stuff, please?

File: src/lib/third_party/src/ndpi_crc.c [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  return crc;
+}
+
+u_int16_t ndpi_crc16_ccit_false(const void *data, size_t n_bytes) {
+  return __crc16(0xFFFF, data, n_bytes);
+}
+
+u_int16_t ndpi_crc16_xmodem(const void *data, size_t n_bytes) {
+  return __crc16(0, data, n_bytes);
+}
```
Comment (IvanNardi) [inline src/lib/third_party/src/ndpi_crc.c]: Could you also extend `fuzz/fuzz_alg_crc32_md5.c` calling these 3 new functions and updating the comment at the beginning of the file, please?
[please, do not change the file name!]

**Review by IvanNardi — CHANGES_REQUESTED** [review]

**Review by IvanNardi — APPROVED** [review]

**Comment by IvanNardi** [comment]
@0xA50C1A1, can I merge it or are you doing to push other stuff?

**Comment by 0xA50C1A1** [comment]
> @0xA50C1A1, can I merge it or are you doing to push other stuff?

Yea, sure.

**Comment by IvanNardi** [comment]
Thanks for all your contributions!

**Pull request merged**